pkg/storegateway/series_refs.go

```diff
@@ -445,3 +453,314 @@ func (s *deduplicatingSeriesChunkRefsSetIterator) Next() bool {
+		l.err = l.from.Err()
+		return false
+	}
+
+	l.currentBatch = l.from.At()
+	err := l.seriesLimiter.Reserve(uint64(l.currentBatch.len()))
+	if err != nil {
+		l.err = errors.Wrap(err, "exceeded series limit")
+		return false
+	}
+
+	var totalChunks int
+	for _, s := range l.currentBatch.series {
+		totalChunks += len(s.chunks)
+	}
+
+	err = l.chunksLimiter.Reserve(uint64(totalChunks))
+	if err != nil {
+		l.err = errors.Wrap(err, "exceeded chunks limit")
+		return false
+	}
+	return true
+}
+
+func (l *limitingSeriesChunkRefsSetIterator) At() seriesChunkRefsSet {
+	return l.currentBatch
+}
+
+func (l *limitingSeriesChunkRefsSetIterator) Err() error {
+	return l.err
+}
+
+type loadingSeriesChunkRefsSetIterator struct {
+	ctx                 context.Context
+	postingsSetIterator *postingsSetsIterator
+	indexr              *bucketIndexReader
+	stats               *safeQueryStats
+	blockID             ulid.ULID
+	shard               *sharding.ShardSelector
+	seriesHasher        seriesHasher
+	skipChunks          bool
+	minTime, maxTime    int64
+
+	symbolizedLsetBuffer []symbolizedLabel
+	chksBuffer           []chunks.Meta
+
+	err        error
+	currentSet seriesChunkRefsSet
+}
+
+func openBlockSeriesChunkRefsSetsIterator(
+	ctx context.Context,
+	batchSize int,
+	indexr *bucketIndexReader, // Index reader for block.
+	blockMeta *metadata.Meta,
+	matchers []*labels.Matcher, // Series matchers.
+	shard *sharding.ShardSelector, // Shard selector.
+	seriesHashCache *hashcache.BlockSeriesHashCache, // Block-specific series hash cache (used only if shard selector is specified).
+	chunksLimiter ChunksLimiter, // Rate limiter for loading chunks.
+	seriesLimiter SeriesLimiter, // Rate limiter for loading series.
+	skipChunks bool, // If true chunks are not loaded and minTime/maxTime are ignored.
+	minTime, maxTime int64, // Series must have data in this time range to be returned (ignored if skipChunks=true).
+	stats *safeQueryStats,
+	logger log.Logger,
+) (seriesChunkRefsSetIterator, error) {
+	if batchSize <= 0 {
+		return nil, errors.New("set size must be a positive number")
+	}
+
+	ps, err := indexr.ExpandedPostings(ctx, matchers, stats)
+	if err != nil {
+		return nil, errors.Wrap(err, "expanded matching postings")
+	}
+
+	// We can't compute the series hash yet because we're still missing the series labels.
+	// However, if the hash is already in the cache, then we can remove all postings for series
+	// not belonging to the shard.
+	if shard != nil {
+		var unsafeStats queryStats
+		ps, unsafeStats = filterPostingsByCachedShardHash(ps, shard, seriesHashCache)
+		stats.merge(&unsafeStats)
+	}
+
+	postingsIterator := newPostingsSetsIterator(ps, batchSize)
+	loadingIterator := newLoadingSeriesChunkRefsSetIterator(
+		ctx,
+		postingsIterator,
+		indexr,
+		stats,
+		blockMeta,
+		shard,
+		cachedSeriesHasher{cache: seriesHashCache},
+		skipChunks,
+		minTime,
+		maxTime,
+	)
+
+	limitingIterator := newLimitingSeriesChunkRefsSetIterator(
```

Comment:
[non blocking] `openBlockSeriesChunkRefsSetsIterator()` is per block. This means that we apply the series limiter for every series and block. If the same series is queried from multiple blocks, then the series is double counted. This is the same behaviour we currently have in `main` so this is not a regression, but with the streaming store-gateway we have the opportunity the limiter up the stack and to actually apply on the merged series from all blocks. We can further discuss it and eventually do it in the working branch.